### PR TITLE
feat(bundles): skip fully unbundled bundles

### PR DESCRIPTION
**Pull request by karlprieb**

### CODE REVIEW

File: src/database/sql/bundles/import.sql [hunk 1]
```diff
@@ -1,43 +1,58 @@
...
 INSERT INTO bundles (
-  id, root_transaction_id, format_id,
-  unbundle_filter_id, index_filter_id,
-  data_item_count, matched_data_item_count,
-  first_queued_at, last_queued_at,
-  first_skipped_at, last_skipped_at,
-  first_unbundled_at, last_unbundled_at,
-  first_fully_indexed_at, last_fully_indexed_at,
-  import_attempt_count, duplicated_data_item_count
+id, root_transaction_id, format_id,
```
Comment (djwhitt) [inline src/database/sql/bundles/import.sql:3]: Nit: let's preserve the original indentation here.

File: src/database/standalone-sqlite.ts [hunk 2]
```diff
@@ -2909,7 +2923,7 @@ export class StandaloneSqliteDatabase
     return this.queueWrite('bundles', 'saveDataItem', [item]);
   }
 
-  saveBundle(bundle: BundleRecord): Promise<void> {
+  async saveBundle(bundle: BundleRecord): Promise<BundleFilterIds> {
```
Comment (dtfiedler) [inline src/database/standalone-sqlite.ts:2926]: nit: if you are not awaiting here you do not need the async. The return type tells caller to expect Promise.

File: src/system.ts [hunk 3]
```diff
@@ -509,13 +509,31 @@ export async function queueBundle(
...
+        previousUnbundleFilterId,
+        previousIndexFilterId,
+      } = await db.saveBundle({
         id: item.id,
         format: 'ans-104',
         unbundleFilter: config.ANS104_UNBUNDLE_FILTER_STRING,
         indexFilter: config.ANS104_INDEX_FILTER_STRING,
         queuedAt: currentUnixTimestamp(),
       });
+
```
[the remark this answers is not in the document]
Comment (djwhitt) [reply src/system.ts:524]: It's not redundant. It's saving different attributes.

File: src/database/sql/bundles/import.sql [hunk 4]
```diff
@@ -1,43 +1,58 @@
...
-  first_unbundled_at = IFNULL(first_unbundled_at, @unbundled_at),
-  last_unbundled_at = IFNULL(@unbundled_at, last_unbundled_at),
-  first_fully_indexed_at = IFNULL(first_fully_indexed_at, @fully_indexed_at),
-  last_fully_indexed_at = @fully_indexed_at,
-  import_attempt_count = CASE
+data_item_count = IFNULL(@data_item_count, data_item_count),
+matched_data_item_count = IFNULL(@matched_data_item_count, matched_data_item_count),
+duplicated_data_item_count = IFNULL(@duplicated_data_item_count, duplicated_data_item_count),
+previous_unbundle_filter_id = CASE
+    WHEN @unbundle_filter_id IS NOT NULL THEN @unbundle_filter_id
```
Comment (djwhitt) [inline src/database/sql/bundles/import.sql:28]: Is this logic correct? It appears this will set `previous_unbundle_filter_id` to the value currently being passed into the SQL statement.

File: src/database/sql/bundles/import.sql [hunk 5]
```diff
@@ -1,43 +1,58 @@
...
+    ELSE bundles.previous_unbundle_filter_id
+END,
+previous_index_filter_id = CASE
+    WHEN @index_filter_id IS NOT NULL THEN @index_filter_id
```
Comment (djwhitt) [inline src/database/sql/bundles/import.sql:32]: Same question here as above.

File: src/database/standalone-sqlite.test.ts [hunk 6]
```diff
@@ -1044,6 +1046,44 @@ describe('StandaloneSqliteDatabase', () => {
       assert.equal(bundlesDb.prepare(sql).get({ id: fromB64Url(id0) }).cnt, 1);
     });
 
+    it('should update previous_unbundle_filter_id when unbundle_filter_id is not nul', async () => {
```
Comment (djwhitt) [inline src/database/standalone-sqlite.test.ts:1049]: Nit: type "nul" instead of "null".

File: src/database/standalone-sqlite.test.ts [hunk 7]
```diff
@@ -1044,6 +1046,44 @@ describe('StandaloneSqliteDatabase', () => {
...
+      });
+
+      previousUnbundleFilterId = bundlesDb
+        .prepare(sql)
+        .get({ id: fromB64Url(id1) }).previous_unbundle_filter_id;
+
+      assert.equal(previousUnbundleFilterId, 1);
+    });
+
+    it('should update previous_index_filter_id when index_filter_id is not nul', async () => {
```
Comment (djwhitt) [inline src/database/standalone-sqlite.test.ts:1068]: Same nit

File: src/database/standalone-sqlite.test.ts [hunk 8]
```diff
@@ -1044,6 +1046,44 @@ describe('StandaloneSqliteDatabase', () => {
...
+        indexFilter: '{"always": true}',
+      });
+
+      previousIndexFilterId = bundlesDb
+        .prepare(sql)
+        .get({ id: fromB64Url(id1) }).previous_index_filter_id;
+
+      assert.equal(previousIndexFilterId, 1);
+    });
+
```
Comment (djwhitt) [inline src/database/standalone-sqlite.test.ts:1086]: We need a test here to check that the previous values are updated correctly when they are not null.

**Review by djwhitt — CHANGES_REQUESTED** [review]
It's not obvious to me this will work as intended. There are no tests for when the previous values were not null.

File: src/database/sql/bundles/import.sql (hunk 4 above)
Comment (karlprieb) [reply src/database/sql/bundles/import.sql]: https://github.com/ar-io/ar-io-node/pull/289/commits/89fabe0f1e734d0bad7129917836224e02bd5454

File: src/database/sql/bundles/import.sql (hunk 1 above)
Comment (karlprieb) [reply src/database/sql/bundles/import.sql]: https://github.com/ar-io/ar-io-node/pull/289/commits/4af080a12cea9ab3e2a6800aeba30d6320aa957a

File: src/database/standalone-sqlite.ts (hunk 2 above)
Comment (karlprieb) [reply src/database/standalone-sqlite.ts]: https://github.com/ar-io/ar-io-node/pull/289/commits/7bed4f5f8cb6345e9dc9d3b22518aae76a315cdf

**Review by djwhitt — APPROVED** [review]
